v2: Generated code fails to compile

The following design causes the generated code to fail compilation
```golang
package design

import (
	. "goa.design/goa/dsl"
)

var _ = Service("Foo", func() {
	Method("list", func() {
		Payload(func() {
			Reference(Foo)
			Attribute("project_id")
			Required("project_id")
		})
		Result(FooList)
		GRPC(func() {})
	})

	Method("subscribe", func() {
		Payload(func() {
			Reference(Foo)
			Attribute("project_id")
			Required("project_id")
		})
		StreamingResult(FooEvent)
		GRPC(func() {})
	})
})

var Foo = ResultType("application/vnd.myapp.foo", func() {
	TypeName("Foo")
	Attributes(func() {
		Field(1, "project_id", UInt, "project_id is the project ID")
		Field(2, "id", String, "id is the Foo ID")
		Field(3, "name", String, "name is the name of Foo")
		Field(4, "description", String, "The description of Foo")
		Required("project_id", "id", "name")
	})

	View("default", func() {
		Attribute("project_id")
		Attribute("id")
		Attribute("name")
	})

	View("expanded", func() {
		Attribute("project_id")
		Attribute("id")
		Attribute("name")
		Attribute("description")
	})
})

var FooList = ResultType("application/vnd.myapp.foo_list", func() {
	TypeName("FooList")
	Attributes(func() {
		Field(1, "count", UInt, "count is the total number of resources in the list.")
		Field(2, "items", CollectionOf(Foo), "items is the list of foo in the list.")
		Required("count")
	})

	View("default", func() {
		Attribute("count")
		Attribute("items")
	})

	View("expanded", func() {
		Attribute("count")
		Attribute("items", func() {
			View("expanded")
		})
	})
})

var FooEvent = ResultType("application/vnd.myapp.foo_event", func() {
	TypeName("FooEvent")
	Attributes(func() {
		Field(1, "op", String, "op is the streaming operation", func() {
			Example("add")
			Enum("add", "change", "remove")
		})
		Field(2, "foo", Foo, "the Foo")
		Required("op", "foo")
	})

	View("default", func() {
		Attribute("op")
		Attribute("foo")
	})

	View("expanded", func() {
		Attribute("op")
		Attribute("foo", func() {
			View("expanded")
		})
	})
})
```

To reproduce the issue, generate the code using
```
goa gen /path/to/design
goa example /path/to/design
```
and then compile the code using
```
go build path/to/cmd/api
```
which causes the following compilation error:
```
gen/grpc/foo/server/types.go:71:14: undefined: ValidateFooCollection
```
I can't exactly reproduce the error I'm seeing in project but this is very similar. In my project the the `ValidateSubscribeReponse` was referring to `ValidateFoo2` function which wasn't generated. In this case, the `ValidateSubscribeResponse` doesn't even validate the `Foo` field.

--- a/codegen/service/example_svc.go
+++ b/codegen/service/example_svc.go
@@ -27,6 +27,9 @@ type (
 		// ResultView is the view to render the result. It is set only if the
 		// result type uses views.
 		ResultView string
+		// StreamInterface is the stream interface in the service package used
+		// by the endpoint implementation.
+		StreamInterface string
 	}
 )
 
@@ -95,6 +98,9 @@ func basicEndpointSection(m *expr.MethodExpr, svcData *Data) *codegen.SectionTem
 			ed.ResultView = view
 		}
 	}
+	if md.ServerStream != nil {
+		ed.StreamInterface = svcData.PkgName + "." + md.ServerStream.Interface
+	}
 	return &codegen.SectionTemplate{
 		Name:   "basic-endpoint",
 		Source: endpointT,
@@ -120,7 +126,7 @@ func New{{ .StructName }}(logger *log.Logger) {{ .PkgName }}.Service {
 	// input: basicEndpointData
 	endpointT = `{{ comment .Description }}
 {{- if .ServerStream }}
-func (s *{{ .ServiceVarName }}srvc) {{ .VarName }}(ctx context.Context{{ if .PayloadFullRef }}, p {{ .PayloadFullRef }}{{ end }}, stream {{ .ServerStream.Interface }}) (err error) {
+func (s *{{ .ServiceVarName }}srvc) {{ .VarName }}(ctx context.Context{{ if .PayloadFullRef }}, p {{ .PayloadFullRef }}{{ end }}, stream {{ .StreamInterface }}) (err error) {
 {{- else }}
 func (s *{{ .ServiceVarName }}srvc) {{ .VarName }}(ctx context.Context{{ if .PayloadFullRef }}, p {{ .PayloadFullRef }}{{ end }}) ({{ if .ResultFullRef }}res {{ .ResultFullRef }}, {{ if .ViewedResult }}{{ if not .ViewedResult.ViewName }}view string, {{ end }}{{ end }} {{ end }}err error) {
 {{- end }}

--- a/codegen/validation.go
+++ b/codegen/validation.go
@@ -314,7 +314,7 @@ func recurseAttribute(ca *ContextualAttribute, nat *expr.NamedAttributeExpr, tar
 			if expr.IsArray(nat.Attribute.Type) {
 				buf.Write(recurseValidationCode(ca, tgt, context, seen).Bytes())
 			} else {
-				if err := userValT.Execute(&buf, map[string]interface{}{"name": Goify(ut.Name(), true), "target": tgt}); err != nil {
+				if err := userValT.Execute(&buf, map[string]interface{}{"name": Goify(ca.Attribute.Name(), true), "target": tgt}); err != nil {
 					panic(err) // bug
 				}
 			}

--- a/grpc/codegen/server.go
+++ b/grpc/codegen/server.go
@@ -197,7 +197,7 @@ type ErrorNamer interface {
 // streamStructTypeT renders the server and client struct types that
 // implements the client and server service stream interfaces.
 // input: StreamData
-const streamStructTypeT = `{{ printf "%s implements the %s.%s interface." .VarName .ServiceInterface | comment }}
+const streamStructTypeT = `{{ printf "%s implements the %s interface." .VarName .ServiceInterface | comment }}
 type {{ .VarName }} struct {
 	stream {{ .Interface }}
 {{- if .Endpoint.Method.ViewedResult }}

--- a/grpc/codegen/service_data.go
+++ b/grpc/codegen/service_data.go
@@ -608,16 +608,14 @@ func collectMessages(at *expr.AttributeExpr, sd *ServiceData, seen map[string]st
 			Ref:         ref,
 			Type:        dt,
 		})
-		ca := protoBufContext(at, "", sd.Scope)
-		if vDef := codegen.RecursiveValidationCode(ca, "message"); vDef != "" {
-			sd.Validations = append(sd.Validations, &ValidationData{
-				Name:    "Validate" + name,
-				Def:     vDef,
-				ArgName: "message",
-				SrcName: name,
-				SrcRef:  ref,
-			})
-		}
+		ca := protoBufContext(att, "", sd.Scope)
+		sd.Validations = append(sd.Validations, &ValidationData{
+			Name:    "Validate" + name,
+			Def:     codegen.RecursiveValidationCode(ca, "message"),
+			ArgName: "message",
+			SrcName: name,
+			SrcRef:  ref,
+		})
 		seen[dt.Name()] = struct{}{}
 		data = append(data, collect(att)...)
 	case *expr.Object:

--- a/grpc/codegen/testdata/streaming_code.go
+++ b/grpc/codegen/testdata/streaming_code.go
@@ -1,7 +1,7 @@
 package testdata
 
 var ServerStreamingServerStructCode = `// MethodServerStreamingUserTypeRPCServerStream implements the
-// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCServerStream.%!s(MISSING)
+// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCServerStream
 // interface.
 type MethodServerStreamingUserTypeRPCServerStream struct {
 	stream pb.ServiceServerStreamingUserTypeRPC_MethodServerStreamingUserTypeRPCServer
@@ -23,7 +23,7 @@ var ServerStreamingServerCloseCode = `func (s *MethodServerStreamingUserTypeRPCS
 `
 
 var ServerStreamingClientStructCode = `// MethodServerStreamingUserTypeRPCClientStream implements the
-// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCClientStream.%!s(MISSING)
+// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCClientStream
 // interface.
 type MethodServerStreamingUserTypeRPCClientStream struct {
 	stream pb.ServiceServerStreamingUserTypeRPC_MethodServerStreamingUserTypeRPCClient
@@ -49,7 +49,7 @@ var ServerStreamingClientCloseCode = `func (s *MethodServerStreamingUserTypeRPCC
 `
 
 var ServerStreamingResultWithViewsServerStructCode = `// MethodServerStreamingUserTypeRPCServerStream implements the
-// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCServerStream.%!s(MISSING)
+// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCServerStream
 // interface.
 type MethodServerStreamingUserTypeRPCServerStream struct {
 	stream pb.ServiceServerStreamingUserTypeRPC_MethodServerStreamingUserTypeRPCServer
@@ -73,7 +73,7 @@ func (s *MethodServerStreamingUserTypeRPCServerStream) SetView(view string) {
 `
 
 var ServerStreamingResultWithViewsClientStructCode = `// MethodServerStreamingUserTypeRPCClientStream implements the
-// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCClientStream.%!s(MISSING)
+// serviceserverstreamingusertyperpc.MethodServerStreamingUserTypeRPCClientStream
 // interface.
 type MethodServerStreamingUserTypeRPCClientStream struct {
 	stream pb.ServiceServerStreamingUserTypeRPC_MethodServerStreamingUserTypeRPCClient
@@ -187,8 +187,7 @@ func (s *MethodServerStreamingMapClientStream) Recv() (map[string]*serviceserver
 `
 
 var ClientStreamingServerStructCode = `// MethodClientStreamingRPCServerStream implements the
-// serviceclientstreamingrpc.MethodClientStreamingRPCServerStream.%!s(MISSING)
-// interface.
+// serviceclientstreamingrpc.MethodClientStreamingRPCServerStream interface.
 type MethodClientStreamingRPCServerStream struct {
 	stream pb.ServiceClientStreamingRPC_MethodClientStreamingRPCServer
 }
@@ -215,8 +214,7 @@ func (s *MethodClientStreamingRPCServerStream) Recv() (int, error) {
 `
 
 var ClientStreamingClientStructCode = `// MethodClientStreamingRPCClientStream implements the
-// serviceclientstreamingrpc.MethodClientStreamingRPCClientStream.%!s(MISSING)
-// interface.
+// serviceclientstreamingrpc.MethodClientStreamingRPCClientStream interface.
 type MethodClientStreamingRPCClientStream struct {
 	stream pb.ServiceClientStreamingRPC_MethodClientStreamingRPCClient
 }
@@ -243,7 +241,7 @@ func (s *MethodClientStreamingRPCClientStream) CloseAndRecv() (string, error) {
 `
 
 var BidirectionalStreamingServerStructCode = `// MethodBidirectionalStreamingRPCServerStream implements the
-// servicebidirectionalstreamingrpc.MethodBidirectionalStreamingRPCServerStream.%!s(MISSING)
+// servicebidirectionalstreamingrpc.MethodBidirectionalStreamingRPCServerStream
 // interface.
 type MethodBidirectionalStreamingRPCServerStream struct {
 	stream pb.ServiceBidirectionalStreamingRPC_MethodBidirectionalStreamingRPCServer
@@ -279,7 +277,7 @@ var BidirectionalStreamingServerCloseCode = `func (s *MethodBidirectionalStreami
 `
 
 var BidirectionalStreamingClientStructCode = `// MethodBidirectionalStreamingRPCClientStream implements the
-// servicebidirectionalstreamingrpc.MethodBidirectionalStreamingRPCClientStream.%!s(MISSING)
+// servicebidirectionalstreamingrpc.MethodBidirectionalStreamingRPCClientStream
 // interface.
 type MethodBidirectionalStreamingRPCClientStream struct {
 	stream pb.ServiceBidirectionalStreamingRPC_MethodBidirectionalStreamingRPCClient

